tests/unit/backend/chalice/api_server/mock_auth.py

```diff
@@ -1,75 +1,97 @@
-
-
-@mock_oauth_app.route("/.well-known/openid-configuration")
-def openid_configuration():
-    data = dict(jwks_uri=f"http://localhost:{PORT}/.well-known/jwks.json")
-    return make_response(jsonify(data))
-
-
-@mock_oauth_app.route("/.well-known/jwks.json")
-def jwks():
-    data = dict(
-        alg="RS256",
-        kty="RSA",
-        use="sig",
-        kid="fake_kid",
-    )
-    return make_response(jsonify(dict(keys=[data])))
-
-
-# function to launch the mock oauth server
-def launch_mock_oauth():
-    mock_oauth_app.run(port=PORT)
+# A mocked out oauth server, which serves all the endpoints needed by the oauth type.
+class MockOauthApp:
+    def __init__(self, port):
+        self.port = port
+
+        # mock flask app
+        self.app = Flask("mock_oauth_app")
+
+        self.app.add_url_rule("/authorize", view_func=self.api_authorize)
+        self.app.add_url_rule("/oauth/token", view_func=self.api_oauth_token, methods=["POST"])
+        self.app.add_url_rule("/v2/logout", view_func=self.api_logout)
+        self.app.add_url_rule("/.well-known/openid-configuration", view_func=self.api_openid_configuration)
+        self.app.add_url_rule("/.well-known/jwks.json", view_func=self.api_jwks)
+
+    def api_authorize(self):
+        callback = request.args.get("redirect_uri")
+        state = request.args.get("state")
+        return redirect(callback + f"?code=fakecode&state={state}")
+
+    def api_oauth_token(self):
+        expires_at = time.time()
+        headers = dict(alg="RS256", kid="fake_kid")
+        payload = dict(
+            name="Fake User", sub="test_user_id", email="fake_user@email.com", email_verified=True, exp=expires_at
+        )
+
+        jwt = jose.jwt.encode(claims=payload, key="mysecret", algorithm="HS256", headers=headers)
+        r = {
+            "access_token": f"access-{time.time()}",
+            "id_token": jwt,
+            "refresh_token": f"random-{time.time()}",
+            "scope": "openid profile email offline",
+            "expires_in": TOKEN_EXPIRES,
+            "token_type": "Bearer",
+            "expires_at": expires_at,
+        }
+        return make_response(jsonify(r))
+
+    def api_logout(self):
+        return_to = request.args.get("returnTo")
+        return redirect(return_to)
+
+    def api_openid_configuration(self):
+        data = dict(jwks_uri=f"http://localhost:{self.port}/.well-known/jwks.json")
+        return make_response(jsonify(data))
+
+    def api_jwks(self):
+        data = dict(
+            alg="RS256",
+            kty="RSA",
+            use="sig",
+            kid="fake_kid",
+        )
+        return make_response(jsonify(dict(keys=[data])))
+
+
+class MockOauthServer:
+    def __init__(self):
+        self.process = None
+        self.port = None
+        self.server_okay = False
+
+    def start(self):
+        self.port = random.randint(10000, 20000)
+        self.process = subprocess.Popen([sys.executable, __file__, str(self.port)])
```

Comment:
Does `multiprocessing.Process` not work here?

Reply:
It should, but I ran into an error when running the full test suite.  I spent a little bit of time 
trying to figure out the issue, but I couldn't spot the problem so I switched to subprocess.

Reply:
Just curious -- do you remember what the error was?